### ISSUE TITLE
Akka HTTP client sends absolute URI in Request-Line

### ISSUE THREAD

**Issue by RichardBradley** [akka/akka#17057]
The Akka HTTP client generates Request-Lines containing an absolute URI, rather than a relative URI as specified by [RFC 7230 sec 5.3.1](https://tools.ietf.org/html/rfc7230#section-5.3):

> When making a request directly to an origin server, other than a
   CONNECT or server-wide OPTIONS request (as detailed below), a client
   MUST send only the absolute path and query components of the target
   URI as the request-target.

Some servers are rejecting requests from my Akka 1.0-M4 app, because it is sending requests with an absolute Request-Line like:

```
PUT http://172.22.145.52:8080/v1/AUTH_1d5faf75c6ec4562ac5bae98d7ea7335/BOSLFuncSpec-c1e6c5dc-1d0e-4df8-bf9a-1e268c343033 HTTP/1.1
```

(I know that the server is in violation of section 5.3.2 "a server MUST accept the absolute-form in requests, even though HTTP/1.1 clients will only send them in requests to proxies.", but I can't change that server.)

**Comment by RichardBradley** [comment]
(This isn't urgent -- I can work around this for now with code like:

```
    request.withUri(request.uri.toRelative)
```
)

**Comment by sirthias** [comment]
Thanks for reporting, Rich! You are of course absolutely right with this one.
The spec is crystal-clear here.

What is especially interesting here is the interaction with another ticket we recently got in spray (https://github.com/spray/spray/issues/1019).

As you correctly point out RFC 7230 demands that a client sends requests to an origin server in what is called "origin-form", i.e. an URI w/o scheme and authority component.
However, https://github.com/spray/spray/issues/1019 describes the problem that spray's client logic doesn't allow requests to URIs like this one

    http://example.com//foo

i.e. ones where the path starts with a double-slash.
This type of URIs is allowed by RFC 3986 in its absolute form. But you cannot simply strip off the scheme and the authority and parse the remainder as a `relative-ref` (as defined by http://tools.ietf.org/html/rfc3986#appendix-A), because relative refs starting with a double slash interpret the first path segment after the double slash as the *authority*.

So, what this means is that the `request-target` of an HTTP request is *not* actually a `relative-ref` as defined by RFC 3986. If one reads the [collected ABNF of RFC 7230](https://tools.ietf.org/html/rfc7230#appendix-B) carefully one actually realizes this.

In summary, all we have to do is to simply really follow RFC 7230 for parsing request targets and not branch out to RFC 3986 a bit too early, as we appear to be doing right now.

//c @jrudolph

**Comment by jrudolph** [comment]
Interesting, indeed. (I had the spray issue on my list of tickets to review because something seemed amiss but didn't get to it yet.)

**Comment by sirthias** [comment]
By the way, this issue is invalid in that the only currently available "HTTP client" provided by Akka is the most basic connection-level infrastructure, which is not supposed to translate absolute URIs and manage `Host` headers.
When you drop down to this level you want full control over the connection, including the question of how exactly the request is to be rendered. We'll need this flexibility for example when opening connection to proxies, for which we *do* need to render absolute request targets.

The feature that you are looking for (automatic management of `Host` headers and request URI translation) is part of the higher-level client-side APIs, which are not merged yet.
(See this PR for a current preview: https://github.com/akka/akka/pull/17036).

Therefore I am closing this ticket here.
(However, thank you very much again for your great feedback, @RichardBradley! Please keep those tickets coming! They are incredibly helpful also in sharpening our own understanding of what exactly we are doing... ;)

**Comment by RichardBradley** [comment]
OK, that sounds correct to me.
I trust there will be a test for this in the final HTTP client :-)
